README.md
# YouTube iframe extension

An extension for [league/commonmark](https://github.com/thephpleague/commonmark)
version 2 built using PHP 8.0. This replaces YouTube links with the embed iframe.

The extension supports for the primary YouTube URL, with and without prefixed with
the `www`. It also supports the short shareable URL using the `youtu.be` domain.

Initially based on the [YouTube extension](https://github.com/zoonru/commonmark-ext-youtube-iframe).

## Installation

The project should be installed via Composer:

```bash
composer require surface/commonmark-ext-youtube-iframe
```

## Usage

Configure your CommonMark `Environment` and add the extension.

```php
use League\CommonMark\Environment\Environment;
use League\CommonMark\MarkdownConverter as Converter;
use Surface\CommonMark\Ext\YouTubeIframe\Extension as YouTubeExtension;

$options = [
    'youtube_iframe' => [
        'width' => 800,
        'height' => 600,
    ]
];

$environment = new Environment($options);
$environment->addExtension(new YouTubeExtension());

$converter = new Converter($environment);

echo $converter->convert('[](https://youtu.be/xxx)');
echo $converter->convert('[](https://www.youtube.com/watch?v=xxx)');
echo $converter->convert('[](https://youtu.be/xxx?height=480)');
echo $converter->convert('[](https://www.youtube.com/watch?v=xxx&width=640)');
```

### Dimensions

You can control the dimensions of the videos by using the `width` and
`height` configuration options.

You can also configure the dimensions using query parameters on the embed URL.
You can provide the `height` or `width` or *both*.

```html
?width=640
?height=480
?width=640&height=480
```

### Full URL

You can disable the parsing of full YouTube URLs (with or without the www) by
using the following option. If this option is disabled, only URLs under the
‘short’ domain will be used (`youtu.be`).

```php
$options = [
    'youtube_iframe' => [
        'full_url' => false,
    ]
];
```

## Testing

There are Unit and Integration tests for the project. These can be run using
the following commands:

```bash
composer test
composer run test
composer run test-unit
composer run test-integration
```

There are also scripts to run code sniffer, mess detector and static analysis:

```bash
composer run sniff
composer run mess
composer run stan
```

## Changelog

Please refer to the [CHANGELOG](CHANGELOG.md) for more information on what has
changed recently.

## License

This library is licensed under the MIT license. See the
[License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.1.0](https://github.com/trovster/commonmark-ext-youtube-iframe/compare/v2.0.0...v2.1.0) (2022-12-06)


### Features

* support configuration options & 'full url' option ([c3a607d](https://github.com/trovster/commonmark-ext-youtube-iframe/commit/c3a607d8ab9052e56436abdd1154ee3b1d12cd1a))

## [2.0.0] - 2022-04-13

### Added

- Update to PHP 8.0 and CommonMark v2.

## [1.0.0] - 2022-04-13

### Added

- Initial release supporting CommonMark v1.


src/Extension.php
<?php

namespace Surface\CommonMark\Ext\YouTubeIframe;

use League\CommonMark\Environment\EnvironmentBuilderInterface;
use League\CommonMark\Event\DocumentParsedEvent;
use League\CommonMark\Extension\ConfigurableExtensionInterface;
use League\CommonMark\Extension\ExtensionInterface;
use League\Config\ConfigurationBuilderInterface;
use Nette\Schema\Expect;
use Surface\CommonMark\Ext\YouTubeIframe\Iframe;
use Surface\CommonMark\Ext\YouTubeIframe\Processor;
use Surface\CommonMark\Ext\YouTubeIframe\Renderer;
use Surface\CommonMark\Ext\YouTubeIframe\Url\Parsers\Full;
use Surface\CommonMark\Ext\YouTubeIframe\Url\Parsers\Short;
use Surface\CommonMark\Ext\YouTubeIframe\Url\Parsers\WithoutWww;

final class Extension implements ExtensionInterface, ConfigurableExtensionInterface
{
    public function register(EnvironmentBuilderInterface $environment): void
    {
        $width = $this->getWidth($environment);
        $height = $this->getHeight($environment);
        $fullScreen = $this->allowFullscreen($environment);
        $processors = $this->getProcessors($environment, $width, $height);

        $environment
            ->addEventListener(DocumentParsedEvent::class, new Processor($processors))
            ->addRenderer(Iframe::class, new Renderer(
                $width,
                $height,
                $fullScreen
            ));
    }

    public function configureSchema(ConfigurationBuilderInterface $builder): void
    {
        $builder->addSchema('youtube_iframe', Expect::structure([
            'width' => Expect::int(800),
            'height' => Expect::int(600),
            'allowfullscreen' => Expect::bool(true),
            'full_url' => Expect::bool(true),
        ]));
    }

    protected function getProcessors(EnvironmentBuilderInterface $environment, int $width, int $height): array
    {
        if (! $this->supportFullUrl($environment)) {
            return [
                new Short($width, $height),
            ];
        }

        return [
            new Short($width, $height),
            new Full($width, $height),
            new WithoutWww($width, $height),
        ];
    }

    protected function supportFullUrl(EnvironmentBuilderInterface $environment): bool
    {
        if ($environment->getConfiguration()->exists('youtube_iframe.full_url')) {
            return (bool) $environment->getConfiguration()->get('youtube_iframe.full_url');
        }

        return true;
    }

    protected function getWidth(EnvironmentBuilderInterface $environment): int
    {
        if ($environment->getConfiguration()->exists('youtube_iframe.width')) {
            return (int) $environment->getConfiguration()->get('youtube_iframe.width');
        }

        return 800;
    }

    protected function getHeight(EnvironmentBuilderInterface $environment): int
    {
        if ($environment->getConfiguration()->exists('youtube_iframe.height')) {
            return (int) $environment->getConfiguration()->get('youtube_iframe.height');
        }

        return 600;
    }

    protected function allowFullscreen(EnvironmentBuilderInterface $environment): bool
    {
        if ($environment->getConfiguration()->exists('youtube_iframe.allowfullscreen')) {
            return (bool) $environment->getConfiguration()->get('youtube_iframe.allowfullscreen');
        }

        return true;
    }
}
